Split up StringTemplate::addClass()

dereuromark opened the pull request:
Replaces https://github.com/cakephp/cakephp/pull/18846 and pulls it into 5.next as BC solution for the time being.
We have then time to think about a 6.x approach/alternative.

But it would be nice if we could already clean this up now, to avoid those PHPStan issues that need silencing or hacking then:
```
  67     Parameter #2 $options of method                                        
         App\View\Widget\DateWidget::formatDateTime() expects array<string, mixed>, 
         array<string, string>|string|null given.        
```

 Added new methods and deprecated the old one
    - addClassToArray() - Type-safe method for adding classes to attribute arrays
    - mergeClasses() - Focused method for merging class lists
    - addClass() - Deprecated (5.3.0, removal in 6.0) with delegation to new methods

dereuromark committed:
Split up StringTemplate::addClass()

dereuromark committed:
Split up StringTemplate::addClass()

othercorey commented on src/View/StringTemplate.php:
```diff
@@ -343,9 +344,54 @@ protected function _formatAttribute(string $key, mixed $value, bool $escape = tr
+    public function mergeClasses(array|string $existing, array|string $new): array
+    {
+        if (!is_array($existing)) {
+            $existing = $existing !== '' ? explode(' ', (string)$existing) : [];
```
Is the (string) cast needed since it must be a string?

dereuromark replied:
We can remove the cast indeed

othercorey commented:
@albertcansado Would you like to comment on this iteration?

dereuromark committed:
Refactor mergeClasses to simplify string handling

dereuromark commented:
@othercorey Did you ping the right persons?

othercorey commented:
Isn't that who commented on the previous PR? Don't know who else has an opinion.

ADmad commented on src/View/StringTemplate.php:
```diff
@@ -343,9 +344,54 @@ protected function _formatAttribute(string $key, mixed $value, bool $escape = tr
+     *
+     * @param array<string>|string $existing The existing class(es).
+     * @param array<string>|string $new The new class(es) to merge.
+     * @return array<string> A unique array of merged classes.
```
Doesn't the old method return string if the 1st arg is string? If so we can retain the same. We can use conditional return type based on argument type which keeps the return type deterministic for static analyzers.

dereuromark replied:
Yeah, thats why the old code needed (array) casting.
These methods indeed are the 6.x code I recommended moving forward.
So those are already "clean" from BC requirements.
For the backport you are right, we probably should consider more BC approach.

Feel free to add a commit, not quite sure how that conditional works.

albertcansado commented:
Hello @othercorey thanks for the mention!

What I can contribute is my own experience/usage, just another one — I’ve replaced those methods with the one I mentioned in the other PR (`classnames`) because it gives me more flexibility and lets me make inline conditional declarations. It saves me from having to write N if statements and call the method to perform the merge (for example). I can declare and merge at the same time.

Also, since it returns a `string`, I can use it outside of `StringTemplate` or expect it to be rendered with the `formatAttributes` method. An example:

```php
<div class="<?= $this->classnames('foo', ['is-open' => $items->count() > 1, 'is-disabled' => !$items->count()]) ?>"></div>
```

But as I said, it’s just a preference

markstory commented on tests/TestCase/View/StringTemplateTest.php:
```diff
@@ -303,123 +303,244 @@ public function testPushPopTemplates(): void
+     */
+    public function testMergeClasses(): void
+    {
+        // Test merging two arrays
+        $result = $this->template->mergeClasses(['existing'], ['new']);
+        $this->assertSame(['existing', 'new'], $result);
```
Why are we adding _two_ new methods? We should pick _one_ way to do this going forward that provides a more typesafe interface that handles several usage scenarios. I think drawing inspiration from [classnames](https://www.npmjs.com/package/classnames) like @albertcansado suggested is an example of an ergonomic interface that could be hard to replicate exactly in typed PHP, but perhaps we can get close. Their docs have the following scenarios outlined:

```javascript
classNames('foo', 'bar'); // => 'foo bar'
classNames('foo', { bar: true }); // => 'foo bar'
classNames({ 'foo-bar': true }); // => 'foo-bar'
classNames({ 'foo-bar': false }); // => ''
classNames({ foo: true }, { bar: true }); // => 'foo bar'
classNames({ foo: true, bar: true }); // => 'foo bar'

// lots of arguments of various types
classNames('foo', { bar: true, duck: false }, 'baz', { quux: true }); // => 'foo bar baz quux'

// other falsy values are just ignored
classNames(null, false, 'bar', undefined, 0, 1, { baz: null }, ''); // => 'bar 1'
```

In PHP I could see this constrained to:

```php
function addClassnames(string $class, array ...$args): string
```

We can't give much type safety for `args` but I think that is on purpose. I think it is important to remember that *most* of the time classname manipulation is done to feed into another helper. Making it easy to dynamically generate class lists from a map of application state is an important scenario for server rendered HTML.

dereuromark replied:
The PR above states the reason for the new IMO correct naming: They reflect what those separate methods now actually do, instead of the god method from before.

The change is fully automated via rector - and also fully BC.

If you have an alternative, go ahead. I don't see a good one at this point, though, that would replace the bad method with smaller ones, while keeping the upgrade efforts minimal.

ADmad replied:
> Why are we adding two new methods?

Currently if you pass `addClass()` an array as the 1st argument, it treat it as a full arguments array and modifies `$input['class']` and returns the array. How will we provide rector based automated updating if we only add new method which returns string?

That said instead of adding 2 new methods we can deprecate the ability to pass string as 1st argument to existing `addClass()` and also deprecate the ability to return a non-array value. So going future it would only deal with modifying a complete attributes array.

Then we add only 1 new method `classNames()`/`addClassNames()` would handles the actual CSS list merging which required API.

dereuromark committed:
Adjust as per review.

dereuromark commented:
I adjusted it per review.

  Final API:

  - addClass() - Works with attributes arrays (deprecated for non-array usage)
  - addClassNames() - Single new method for CSS class merging
  
  I dont like it compared to the API proposed, it is less speaking IMO and more confusing long term.
  But if that at least solves the issues for now, fair enough.

markstory commented on tests/TestCase/View/StringTemplateTest.php:
```diff
@@ -304,58 +304,148 @@ public function testPushPopTemplates(): void
+    public function testAddClassNames(): void
+    {
+        // Test merging two arrays
+        $result = $this->template->addClassNames(['existing'], ['new']);
+        $this->assertSame(['existing', 'new'], $result);
+
+        // Test merging string with array
+        $result = $this->template->addClassNames('existing', ['new']);
+        $this->assertSame(['existing', 'new'], $result);
```
I think this is going in a better direction from an ergonomics point of view. I'll try to make sometime this weekend to try out my suggestion as well.

dereuromark commented:
Should we first merge https://github.com/cakephp/cakephp/pull/19037 and then base this on top of it?

markstory commented:
> Should we first merge https://github.com/cakephp/cakephp/pull/19037 and then base this on top of it?

If you're ok with 19037, lets get it merged and then we can figure out which methods we should keep. If we have more than one use case we can support more input types. I think we can get the test coverage good enough to support multiple scenarios.

markstory committed:
Add more ergonomic map input for classnames (#19037)

* Add more ergonomic map input for classnames

This makes it much simpler to pass entity 'state' into classnames and
drive styling and interaction hooks.

Refs #18999

* Apply suggestion

Co-authored-by: ADmad <admad.coder@gmail.com>

---------

Co-authored-by: Mark Scherer <dereuromark@users.noreply.github.com>
Co-authored-by: ADmad <admad.coder@gmail.com>

dereuromark committed:
Merge branch '5.next' into 5.next-string-template

ADmad commented on src/View/StringTemplate.php:
```diff
@@ -344,45 +345,102 @@ protected function _formatAttribute(string $key, mixed $value, bool $escape = tr
+        }
+        if (!is_array($new)) {
+            $new = $new !== '' ? explode(' ', $new) : [];
+        }
```
```suggestion
        if (is_string($existing)) {
            $existing = $existing === '' ? [] : explode(' ', $existing);
        }
        if (is_string($new)) {
            $new = $new === '' ? [], explode(' ', $new);
        }
```
Avoid negation.

ADmad committed:
Avoid negation
